beancount-parser/src/error.rs

```diff
@@ -0,0 +1,186 @@
+use std::error::Error;
+use std::fmt;
+
+use pest::Span;
+
+use super::Rule;
+
+/// Details about a `BasicParseError`
+#[derive(Clone, Debug, PartialEq)]
+pub enum ParseErrorKind {
+    /// Input is invalid in some way
+    InvalidInput { message: String },
+    /// Parser has reached an invalid state (most likely a bug in the parser)
+    InvalidParserState { message: String },
+}
```

Comment:
Below, we have a decimal parse error kind, but it is wrapped into the `InvalidInput` variant.  What are your thoughts on instead having an explicit `InvalidDecimal { message: String }` variant here?  We could still have a catch-all `InvalidInput` variant for any parse errors that don't fit into more explicit buckets.  This would give users more flexibility in their reaction to errors, e.g. "I want to explicitly handle the case for invalid decimals differently than other more general invalid input".

I personally like how Servo does this with its `StyleParseErrorKind`.  Note the other `InvalidValue` catch-all.  We probably won't have as many error types as Servo, so perhaps this is less important, but still could be useful for our users.

https://github.com/servo/servo/blob/30ca50ae985fca0c7f6a97bdad05921bb411cd3c/components/style_traits/lib.rs#L111#L170

I think `InvalidParserState` is also a good one to keep.

Reply:
Yeah, having a separate `InvalidDecimal` error kind would probably be a good idea, since `rust_decimal` has limitations on the numbers it can represent.

Reply:
For `InvalidInput` I think a `&'static str` would work as well, that removes the need to allocate.

It would also be possible to directly expose the underlying decimal error and derive a `#[from]`, but I'm not sure that's desirable. It would become part of the public API, although `Decimal` already is.

Reply:
The message in `InvalidInput` is generated in the `From<pest::error::Error>` impl so can't see how it can be `'static`.

I didn't go for a `From<rust_decimal::Error> for ParseError` impl since that would mean that I don't get any span information. But I have added the decimal error as `source` on the `ParseError` like I do for pest errors.

Reply:
Ah, perfect! I might be wrong, but I think all values for the message come directly from the match statement, right? But it's fine either way.